File: cplusplus/combinationSum.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> out;

        sort(candidates.begin(), candidates.end());

        combinSumHelper(ans, out, candidates, target, 0);

        return ans;        
    }

    void combinSumHelper(vector<vector<int>> &ans, vector<int> &out, vector<int> &candidates, int target, int start) {
        if (target == 0) {
            ans.push_back(out);
            return;
        }
        if (target < 0) {
            return;
        }

        for (int i=start; i<candidates.size(); i++) {
            if (i > start && candidates[i] == candidates[i-1]) {
                continue;
            }

            out.push_back(candidates[i]);
            combinSumHelper(ans, out, candidates, target-candidates[i], i+1);
            out.pop_back();
        }
    }
};

File: go/combinationSum.go
package main

import (
	"sort"
)

func combinationSum1(candidates []int, target int) [][]int {
	ans := [][]int{}

	sort.Ints(candidates)  // nedd it? yes, otherwise it could not avoid repeating

	combinationSumHelper1(ans, []int{}, candidates, target, 0, 0)

	return ans
}

func combinationSumHelper1(ans [][]int, oneAns []int, candidates []int, target int, sum int, start int) {
	if sum == target {
		ans = append(ans, oneAns)
		return
	}
	if sum > target {
		return
	}

	for i:=start; i<len(candidates); i++ {
		c := candidates[i]
		newAns := make([]int, len(oneAns))
		copy(newAns, oneAns)
		newAns = append(newAns, c) 
		combinationSumHelper1(ans, newAns, candidates, target, sum+c, i)
		newAns = newAns[:len(newAns)-1]
	}
}